pkg/monitortestlibrary/podaccess/pod_logs_stream_all.go

```diff
@@ -0,0 +1,252 @@
+	c.removeAllWatchers(ctx)
+}
+
+func (c *PodsStreamer) syncPods(ctx context.Context) error {
+	pods, err := c.podLister.Pods(c.namespaceName).List(c.selector)
+	if err != nil {
+		return err
+	}
+
+	watchersForCurrPods := map[podKey]*watcher{}
+	for i := range pods {
+		pod := pods[i]
+		// skip pods that are not on nodes
+		if len(pod.Spec.NodeName) == 0 {
+			continue
+		}
+
+		currKey := podKey{
+			namespace: pod.Namespace,
+			name:      pod.Name,
+		}
+		watchersForCurrPods[currKey] = &watcher{
+			pod:           pod,
+			containerName: c.containerName,
+		}
+	}
+	if len(watchersForCurrPods) == 0 {
+		c.removeAllWatchers(ctx)
+		return nil
+	}
+
+	c.watcherLock.Lock()
+	defer c.watcherLock.Unlock()
+
+	// watchersToDelete holds all the watchers present in the running controller and not needed based on the endpointslice state.
+	watchersToDelete := map[podKey]*watcher{}
+	for watcherKey := range c.watchers {
+		knownWatcher := c.watchers[watcherKey]
+		if _, ok := watchersForCurrPods[watcherKey]; !ok {
+			watchersToDelete[watcherKey] = knownWatcher
+		}
+	}
+	for watcherKey, watcherToDelete := range watchersToDelete {
+		watcherToDelete.podStreamer.Stop(ctx)
+		delete(c.watchers, watcherKey)
+	}
+
+	for watcherKey := range watchersForCurrPods {
+		if _, ok := c.watchers[watcherKey]; ok {
+			continue
+		}
+		newWatcher := watchersForCurrPods[watcherKey]
+		newWatcher.podStreamer = NewPodStreamer(c.kubeClient, watcherKey.namespace, watcherKey.name, c.containerName)
+		go newWatcher.podStreamer.Run(ctx)
+
+		in, inErr := newWatcher.podStreamer.Output()
+		go c.forwardLogLines(ctx, in)
+		go c.forwardLogErrors(ctx, inErr)
+
+		c.watchers[watcherKey] = newWatcher
+	}
+
+	return err
+}
+
+// forwardLogLines exits when ctx is done or in is finished and we've consumed everything
+func (c *PodsStreamer) forwardLogLines(ctx context.Context, in chan LogLineContent) {
+	defer utilruntime.HandleCrash()
+
+	for {
+		select {
+		case curr, ok := <-in:
+			if !ok {
+				return
+			}
+			c.logHandler.HandleLogLine(curr)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *PodsStreamer) forwardLogErrors(ctx context.Context, inErr chan LogError) {
+	defer utilruntime.HandleCrash()
+
+	for {
+		select {
+		case curr, ok := <-inErr:
+			if !ok {
+				return
+			}
+			klog.Error(curr)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *PodsStreamer) removeAllWatchers(ctx context.Context) {
+	c.watcherLock.Lock()
+	defer c.watcherLock.Unlock()
+
+	wg := sync.WaitGroup{}
+	for i := range c.watchers {
+		watcherToDelete := c.watchers[i]
+		wg.Add(1)
+
+		// these can take a bit to shutdown, shut them down in parallel.
+		go func(ctx context.Context, watcherToDelete *watcher) {
+			defer utilruntime.HandleCrash()
+			defer wg.Done()
+			watcherToDelete.podStreamer.Stop(ctx)
+		}(ctx, watcherToDelete)
+	}
+	wg.Wait()
+
+	c.watchers = map[podKey]*watcher{}
+}
+
+// Run starts the controller and blocks until stopCh is closed.
```

Comment:
Nit: I realize this is probably boilerplate but `stopCh` here refers to ctx>Done() right? Thought it was finishedCleanup but that's just the signal for a clean shutdown I think.